docs/reference/configuration.rst
=============================
Project configuration options
=============================

Briefcase is a `PEP518 <https://peps.python.org/pep-0518/>`__-compliant build
tool. It uses a ``pyproject.toml`` file, in the root directory of your project,
to provide build instructions for the packaged file.

If you have an application called "My App", with source code in the ``src/myapp``
directory, the simplest possible ``pyproject.toml`` Briefcase configuration
file would be::

    [tool.briefcase]
    project_name = "My Project"
    bundle = "com.example"
    version = "0.1"

    [tool.briefcase.app.myapp]
    formal_name = "My App"
    description = "My first Briefcase App"
    sources = ['src/myapp']

The configuration sections are tool specific, and start with the prefix
``tool.briefcase``.

The location of the ``pyproject.toml`` file is treated as the root of the
project definition. Briefcase should be invoked in a directory that contains a
``pyproject.toml`` file, and all relative file path references contained in the
``pyproject.toml`` file will be interpreted relative to the directory that
contains the ``pyproject.toml`` file.

Configuration sections
======================

A project that is packaged by Briefcase can declare multiple *applications*.
Each application is a distributable product of the build process. A simple
project will only have a single application. However, a complex project may
contain multiple applications with shared code.

Each setting can be specified:

* At the level of an output format (e.g., settings specific to building macOS
  DMGs);
* At the level of an platform for an app (e.g., macOS specific settings);
* At the level of an individual app; or
* Globally, for all applications in the project.

When building an application in a particular output format, Briefcase will look
for settings in the same order. For example, if you're building a macOS DMG for
an application called ``myapp``, Briefcase will look for macOS DMG settings for
``myapp``, then for macOS settings for ``myapp``, then for ``myapp`` settings,
then for project-level settings.

``[tool.briefcase]``
--------------------

The base ``[tool.briefcase]`` section declares settings that project specific,
or are are common to all applications in this repository.

``[tool.briefcase.app.<app name>]``
-----------------------------------

Configuration options for a specific application.

``<app name>`` must adhere to a valid Python distribution name as specified in
`PEP508 <https://peps.python.org/pep-0508/#names>`__. The app name must also
*not* be a reserved word in Python, Java or JavaScript (i.e., app names like
``switch`` or ``pass`` would not be valid); and it may not include any of the
`filenames prohibited by Windows
<https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions>`__
(i.e., ``CON``, ``PRN``, or ``LPT1``).

``[tool.briefcase.app.<app name>.<platform>]``
----------------------------------------------

Configuration options for an application that are platform specific. The
platform must match a name for a platform supported by Briefcase (e.g.,
``macOS`` or ``windows``). A list of the platforms supported by Briefcase can
be obtained by running ``briefcase -h``, and inspecting the help for the
``platform`` option

``[tool.briefcase.app.<app name>.<platform>.<output format>]``
--------------------------------------------------------------

Configuration options that are specific to a particular output format. For
example, macOS applications can be generated in ``app`` or ``dmg`` format.

This section can contain additional layers. for example, an app targeting the
Linux ``system`` backend can define a ``tool.briefcase.app.<app
name>.linux.system.ubuntu.jammy`` section to provide configurations specific to
Ubuntu 22.04 "Jammy" deployments. See the documentation for each backend for
more details.

Project configuration
=====================

Required values
---------------

``bundle``
~~~~~~~~~~

A reverse-domain name that can be used to identify resources for the
application e.g., ``com.example``. The bundle identifier will be combined with
the app name to produce a unique application identifier - e.g., if the bundle
identifier is ``com.example`` and the app name is ``myapp``, the application
will be identified as ``com.example.myapp``.

``project_name``
~~~~~~~~~~~~~~~~

The project is the collection of all applications that are described by the
briefcase configuration. For projects with a single app, this may be the same
as the formal name of the solitary packaged app.

``version``
~~~~~~~~~~~

A `PEP440 <https://peps.python.org/pep-0440/>`__ compliant version string.

Examples of valid version strings:

* ``1.0``
* ``1.2.3``
* ``1.2.3.dev4`` - A development release
* ``1.2.3a5`` - An alpha pre-release
* ``1.2.3b6`` - A Beta pre-release
* ``1.2.3rc7`` - A release candidate
* ``1.2.3.post8`` - A post-release

Optional values
---------------

``author``
~~~~~~~~~~

The person or organization responsible for the project.

``author_email``
~~~~~~~~~~~~~~~~

The contact email address for the person or organization responsible for the
project.

``url``
~~~~~~~

A URL where more details about the project can be found.

Application configuration
=========================

Required
--------

``description``
~~~~~~~~~~~~~~~

A short, one-line description of the purpose of the application.

``sources``
~~~~~~~~~~~

A list of paths, relative to the ``pyproject.toml`` file, where source code for
the application can be found. The contents of any named files or folders will be
copied into the application bundle. Parent directories in any named path will
not be included. For example, if you specify ``src/myapp`` as a source, the
contents of the ``myapp`` folder will be copied into the application bundle; the
``src`` directory will not be reproduced.

Unlike most other keys in a configuration file, ``sources`` is a *cumulative*
setting. If an application defines sources at the global level, application
level, *and* platform level, the final set of sources will be the
*concatenation* of sources from all levels, starting from least to most
specific.

Optional values
---------------

``accent_color``
~~~~~~~~~~~~~~~~

A hexadecimal RGB color value (e.g., ``#D81B60``) for a subtle secondary color
to be used throughout an application to call attention to key elements. This
setting is only used if the platform allows color modification, otherwise it
is ignored.

``author``
~~~~~~~~~~

The person or organization responsible for the application.

``author_email``
~~~~~~~~~~~~~~~~

The contact email address for the person or organization responsible for the
application.

``build``
~~~~~~~~~

A build identifier. An integer, used in addition to the version specifier,
to identify a specific compiled version of an application.

``cleanup_paths``
~~~~~~~~~~~~~~~~~

A list of strings describing paths that will be *removed* from the project after
the installation of the support package and app code. The paths provided will be
interpreted relative to the app bundle folder (e.g., the ``macOS/app/My App``
folder in the case of a macOS app).

Paths can be:
 * An explicit reference to a single file
 * An explicit reference to a single directory
 * Any file system glob accepted by ``pathlib.glob`` (See `the Python
   documentation for details
   <https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob>`__)

Paths are treated as format strings prior to glob expansion. You can use Python
string formatting to include references to configuration properties of the app
(e.g., ``app.formal_name``, ``app.version``, etc).

For example, the following ``cleanup_paths`` specification::

    cleanup_paths = [
        "path/to/unneeded_file.txt",
        "path/to/unneeded_directory",
        "path/**/*.exe",
        "{app.formal_name}/content/extra.doc"
    ]

on an app with a formal name of "My App" would remove:

1. The file ``path/to/unneeded_file.txt``
2. The directory ``path/to/unneeded_directory``
3. Any ``.exe`` file in ``path`` or its subdirectories.
4. The file ``My App/content/extra.doc``.

``console_app``
~~~~~~~~~~~~~~~

A Boolean describing if the app is a console app, or a GUI app. Defaults to ``False``
(producing a GUI app). This setting has no effect on platforms that do not support
a console mode (e.g., web or mobile platforms).

``exit_regex``
~~~~~~~~~~~~~~

A regular expression that will be executed against the console output generated
by an application. If/when the regular expression find match, the application
will be terminated; the line matching the regular expression will *not* be
output to the console. Used by Briefcase to monitor test suites; however, the
filter will also be honored on normal ``run`` invocations.

The regular expression should capture a single group named ``returncode``,
capturing the integer exit status that should be reported for the process. The
default value for this regular expression is ``^>>>>>>>>>> EXIT
(?P<returncode>.*) <<<<<<<<<<$`` The regex will be compiled with the
``re.MULTILINE`` flag enabled.

``formal_name``
~~~~~~~~~~~~~~~

The application name as it should be displayed to humans. This name may contain
capitalization and punctuation. If it is not specified, the ``name`` will be
used.

``icon``
~~~~~~~~

A path, relative to the directory where the ``pyproject.toml`` file is located,
to an image to use as the icon for the application. The path should *exclude*
the extension; Briefcase will append a platform appropriate extension when
configuring the application. For example, an icon specification of ``icon =
"resources/icon"`` will use ``resources/icon.icns`` on macOS, and
``resources/icon.ico`` on Windows.

Some platforms require multiple icons, at different sizes; these will be
handled by appending the required size to the provided icon name. For example,
iOS requires multiple icon sizes (ranging from 20px to 1024px); Briefcase will
look for ``resources/icon-20.png``, ``resources/icon-1024.png``, and so on. The
sizes that are required are determined by the platform template.

``installer_icon``
~~~~~~~~~~~~~~~~~~

A path, relative to the directory where the ``pyproject.toml`` file is located,
to an image to use as the icon for the installer. As with ``icon``, the
path should *exclude* the extension, and a platform-appropriate extension will
be appended when the application is built.

``installer_background``
~~~~~~~~~~~~~~~~~~~~~~~~

A path, relative to the directory where the ``pyproject.toml`` file is located,
to an image to use as the background for the installer. The path should
*exclude* the extension, and a platform-appropriate extension will be appended
when the application is built.

``long_description``
~~~~~~~~~~~~~~~~~~~~

A longer description of the purpose of the application. This description can be
multiple paragraphs, if necessary. The long description *must not* be a copy of
the ``description``, or include the ``description`` as the first line of the
``long_description``.

``primary_color``
~~~~~~~~~~~~~~~~~

A hexadecimal RGB color value (e.g., ``#008577``) to use as the primary color
for the application. This setting is only used if the platform allows color
modification, otherwise it is ignored.

``primary_color_dark``
~~~~~~~~~~~~~~~~~~~~~~

A hexadecimal RGB color value (e.g., ``#008577``) used alongside the primary
color. This setting is only used if the platform allows color modification,
otherwise it is ignored.


.. _configuration-requires-key:

``requires``
~~~~~~~~~~~~

A list of packages that must be packaged with this application.

Unlike most other keys in a configuration file, ``requires`` is a *cumulative*
setting. If an application defines requirements at the global level,
application level, *and* platform level, the final set of requirements will be
the *concatenation* of requirements from all levels, starting from least to
most specific.

Any PEP 508 version specifier is legal. For example:

* Bare package name::

    requires = ["pillow"]

* Package name with version specifier::

    requires = ["pillow==9.1.0"]

* Install from source using the ``--no-binary`` entry::

    requires = [
        "pillow==9.1.0",
        "--no-binary", "pillow",
    ]

* Git repository::

    requires=["git+https://github.com/beeware/briefcase.git"]

* Local directory::

    requires=["mysrc/myapp"]

* Local wheel file::

    requires=["fullpath/wheelfile.whl"]

``revision``
~~~~~~~~~~~~

An identifier used to differentiate specific builds of the same version of an
app. Defaults to ``1`` if not provided.

``splash_background_color``
~~~~~~~~~~~~~~~~~~~~~~~~~~~

A hexadecimal RGB color value (e.g., ``#6495ED``) to use as the background
color for splash screens.

If the platform output format does not use a splash screen, this setting is
ignored.

``support_package``
~~~~~~~~~~~~~~~~~~~

A file path or URL pointing at a tarball containing a Python support package.
(i.e., a precompiled, embeddable Python interpreter for the platform)

If this setting is not provided, Briefcase will use the default support
package for the platform.

``support_revision``
~~~~~~~~~~~~~~~~~~~~

The specific revision of a support package that should be used. By default,
Briefcase will use the support package revision nominated by the application
template. If you specify a support revision, that will override the revision
nominated by the application template.

If you specify an explicit support package (either as a URL or a file path),
this argument is ignored.

``supported``
~~~~~~~~~~~~~

Indicates that the platform is not supported. For example, if you know that
the app cannot be deployed to Android for some reason, you can explicitly
prevent deployment by setting ``supported=False`` in the Android section of
the app configuration file.

If ``supported`` is set to ``false``, the create command will fail, advising
the user of the limitation.

``template``
~~~~~~~~~~~~

A file path or URL pointing at a `cookiecutter
<https://github.com/cookiecutter/cookiecutter>`__ template for the output
format.

If this setting is not provided, Briefcase will use a default template for
the output format and Python version.

``template_branch``
~~~~~~~~~~~~~~~~~~~

The branch of the project template to use when generating the app. If the
template is a local file, this attribute will be ignored. If not specified,
Briefcase will use a branch matching the version of Briefcase that is being
used (i.e., if you're using Briefcase 0.3.9, Briefcase will use the
``v0.3.9`` template branch when generating the app). If you're using a
development version of Briefcase, Briefcase will use the ``main`` branch of the
template.

``test_requires``
~~~~~~~~~~~~~~~~~

A list of packages that are required for the test suite to run.

Unlike most other keys in a configuration file, ``test_requires`` is a
*cumulative* setting. If an application defines requirements at the global
level, application level, *and* platform level, the final set of requirements
will be the *concatenation* of requirements from all levels, starting from least
to most specific.

See :ref:`requires <configuration-requires-key>` for examples.

``test_sources``
~~~~~~~~~~~~~~~~

A list of paths, relative to the ``pyproject.toml`` file, where test code for
the application can be found. The contents of any named files or folders will be
copied into the application bundle. Parent directories in any named path will
not be included. For example, if you specify ``src/myapp`` as a source, the
contents of the ``myapp`` folder will be copied into the application bundle; the

src/briefcase/config.py
import copy
import json
import keyword
import re
import sys
import unicodedata
from types import SimpleNamespace

if sys.version_info >= (3, 11):  # pragma: no-cover-if-lt-py311
    import tomllib
else:  # pragma: no-cover-if-gte-py311
    import tomli as tomllib

from briefcase.platforms import get_output_formats, get_platforms

from .constants import RESERVED_WORDS
from .exceptions import BriefcaseConfigError

# PEP 508 restricts the naming of modules. The PEP defines a regex that uses
# re.IGNORECASE; but in in practice, packaging uses a version that rolls out the lower
# case, which has very slightly different semantics with non-ASCII characters. This
# definition is the one from
# https://github.com/pypa/packaging/blob/24.0/src/packaging/_tokenizer.py#L80
PEP508_NAME_RE = re.compile(r"^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9._-]*[a-zA-Z0-9])$")


def is_valid_pep508_name(app_name):
    """Determine if the name is valid by PEP508 rules."""
    return PEP508_NAME_RE.match(app_name)


def is_reserved_keyword(app_name):
    """Determine if the name is a reserved keyword."""
    return keyword.iskeyword(app_name.lower()) or app_name.lower() in RESERVED_WORDS


def is_valid_app_name(app_name):
    return not is_reserved_keyword(app_name) and is_valid_pep508_name(app_name)


def make_class_name(formal_name):
    """Construct a valid class name from a formal name.

    :param formal_name: The formal name
    :returns: The app's class name
    """
    # Identifiers (including class names) can be unicode.
    # https://docs.python.org/3/reference/lexical_analysis.html#identifiers
    xid_start = {
        "Lu",  # uppercase letters
        "Ll",  # lowercase letters
        "Lt",  # titlecase letters
        "Lm",  # modifier letters
        "Lo",  # other letters
        "Nl",  # letter numbers
    }
    xid_continue = xid_start.union(
        {
            "Mn",  # nonspacing marks
            "Mc",  # spacing combining marks
            "Nd",  # decimal number
            "Pc",  # connector punctuations
        }
    )

    # Normalize to NFKC form, then remove any character that isn't
    # in the allowed categories, or is the underscore character;
    # Capitalize the resulting word.
    class_name = "".join(
        ch
        for ch in unicodedata.normalize("NFKC", formal_name)
        if unicodedata.category(ch) in xid_continue or ch in {"_"}
    )

    # If the first character isn't in the 'start' character set,
    # and it isn't already an underscore, prepend an underscore.
    if (
        class_name
        and unicodedata.category(class_name[0]) not in xid_start
        and class_name[0] != "_"
    ):
        class_name = f"_{class_name}"

    return class_name


VALID_BUNDLE_RE = re.compile(r"[a-zA-Z0-9-]+(\.[a-zA-Z0-9-]+)+$")


def is_valid_bundle_identifier(bundle):
    # Ensure the bundle identifier follows the basi
    if not VALID_BUNDLE_RE.match(bundle):
        return False

    return True


# This is the canonical definition from PEP440, modified to include named groups
PEP440_CANONICAL_VERSION_PATTERN_RE = re.compile(
    r"^((?P<epoch>[1-9][0-9]*)!)?"
    r"(?P<release>(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*)"
    r"((?P<pre_tag>a|b|rc)(?P<pre_value>0|[1-9][0-9]*))?"
    r"(\.post(?P<post>0|[1-9][0-9]*))?"
    r"(\.dev(?P<dev>0|[1-9][0-9]*))?$"
)


def is_pep440_canonical_version(version):
    """Determine if the string describes a valid PEP440 canonical version specifier.

    This implementation comes directly from PEP440 itself.

    :returns: True if the version string is valid; false otherwise.
    """
    return PEP440_CANONICAL_VERSION_PATTERN_RE.match(version) is not None


def parsed_version(version):
    """Return a parsed version string.

    :param version: The parsed version string
    """
    groupdict = PEP440_CANONICAL_VERSION_PATTERN_RE.match(version).groupdict()

    # Convert dot separated string of integers to tuple of integers
    groupdict["release"] = tuple(int(p) for p in groupdict.pop("release").split("."))

    # Convert strings to values
    for key in ("epoch", "pre_value", "post", "dev"):
        try:
            groupdict[key] = int(groupdict[key])
        except TypeError:
            pass

    tag = groupdict.pop("pre_tag")
    value = groupdict.pop("pre_value")
    groupdict["pre"] = (tag, value) if tag is not None else None
    return SimpleNamespace(**groupdict)


class BaseConfig:
    def __init__(self, **kwargs):
        for attr, value in kwargs.items():
            setattr(self, attr, value)

    def to_fancy_string(self) -> str:
        return f"{self.__repr__()}\n{json.dumps(vars(self), indent=4)}"


class GlobalConfig(BaseConfig):
    def __init__(
        self,
        project_name,
        version,
        bundle,
        license=None,
        url=None,
        author=None,
        author_email=None,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self.project_name = project_name
        self.version = version
        self.bundle = bundle
        self.url = url
        self.author = author
        self.author_email = author_email
        self.license = license

        # Version number is PEP440 compliant:
        if not is_pep440_canonical_version(self.version):
            raise BriefcaseConfigError(
                f"Version number ({self.version}) is not valid.\n\n"
                "Version numbers must be PEP440 compliant; "
                "see https://www.python.org/dev/peps/pep-0440/ for details."
            )

    def __repr__(self):
        return f"<{self.project_name} v{self.version} GlobalConfig>"


class AppConfig(BaseConfig):
    def __init__(
        self,
        app_name,
        version,
        bundle,
        description,
        sources,
        license,
        formal_name=None,
        url=None,
        author=None,
        author_email=None,
        requires=None,
        icon=None,
        document_type=None,
        permission=None,
        template=None,
        template_branch=None,
        test_sources=None,
        test_requires=None,
        supported=True,
        long_description=None,
        console_app=False,
        skip_signing=False,
        **kwargs,
    ):
        super().__init__(**kwargs)

        # All app configs are created in unfinalized draft form.
        self.__draft__ = True

        self.app_name = app_name
        self.version = version
        self.bundle = bundle
        # Description can only be a single line. Ignore everything else.
        self.description = description.split("\n")[0]
        self.sources = sources
        self.formal_name = app_name if formal_name is None else formal_name
        self.url = url
        self.author = author
        self.author_email = author_email
        self.requires = requires
        self.icon = icon
        self.document_types = {} if document_type is None else document_type
        self.permission = {} if permission is None else permission
        self.template = template
        self.template_branch = template_branch
        self.test_sources = test_sources
        self.test_requires = test_requires
        self.supported = supported
        self.long_description = long_description
        self.license = license
        self.console_app = console_app
        self.skip_signing = skip_signing

        if not is_valid_app_name(self.app_name):
            raise BriefcaseConfigError(
                f"{self.app_name!r} is not a valid app name.\n\n"
                "App names must not be reserved keywords such as 'and', 'for' and 'while'.\n"
                "They must also be PEP508 compliant (i.e., they can only include letters,\n"
                "numbers, '-' and '_'; must start with a letter; and cannot end with '-' or '_')."
            )

        if not is_valid_bundle_identifier(self.bundle):
            raise BriefcaseConfigError(
                f"{self.bundle!r} is not a valid bundle identifier.\n\n"
                "The bundle should be a reversed domain name. It must contain at least 2\n"
                "dot-separated sections; each section may only include letters, numbers,\n"
                "and hyphens; and each section may not contain any reserved words (like\n"
                "'switch', or 'while')."
            )

        # Version number is PEP440 compliant:
        if not is_pep440_canonical_version(self.version):
            raise BriefcaseConfigError(
                f"Version number for {self.app_name!r} ({self.version}) is not valid.\n\n"
                "Version numbers must be PEP440 compliant; "
                "see https://www.python.org/dev/peps/pep-0440/ for details."
            )

        # Sources list doesn't include any duplicates
        source_modules = {source.rsplit("/", 1)[-1] for source in self.sources}
        if len(self.sources) != len(source_modules):
            raise BriefcaseConfigError(
                f"The `sources` list for {self.app_name!r} contains duplicated "
                "package names."
            )

        # There is, at least, a source for the app module
        if self.module_name not in source_modules:
            raise BriefcaseConfigError(
                f"The `sources` list for {self.app_name!r} does not include a "
                f"package named {self.module_name!r}."
            )

    def __repr__(self):
        return f"<{self.bundle_identifier} v{self.version} AppConfig>"

    @property
    def module_name(self):
        """The module name for the app.

        This is derived from the name, but:
        * all `-` have been replaced with `_`.
        """
        return self.app_name.replace("-", "_")

    @property
    def bundle_name(self):
        """The bundle name for the app.

        This is derived from the app name, but:
        * all `_` have been replaced with `-`.
        """
        return self.app_name.replace("_", "-")

    @property
    def bundle_identifier(self):
        """The bundle identifier for the app.

        This is derived from the bundle and the bundle name, joined by a `.`.
        """
        return f"{self.bundle}.{self.bundle_name}"

    @property
    def class_name(self):
        """The class name for the app.

        This is derived from the formal name for the app.
        """
        return make_class_name(self.formal_name)

    @property
    def package_name(self):
        """The bundle name of the app, with `-` replaced with `_` to create something
        that can be used a namespace identifier on Python or Java, similar to
        `module_name`."""
        return self.bundle.replace("-", "_")

    def PYTHONPATH(self, test_mode):
        """The PYTHONPATH modifications needed to run this app.

        :param test_mode: Should test_mode sources be included?
        """
        paths = []
        sources = self.sources
        if test_mode and self.test_sources:
            sources.extend(self.test_sources)

        for source in sources:
            path = "/".join(source.rsplit("/", 1)[:-1])
            if path not in paths:
                paths.append(path)
        return paths

    def main_module(self, test_mode: bool):
        """The path to the main module for the app.

        In normal operation, this is ``app.module_name``; however,
        in test mode, it is prefixed with ``tests.``.

        :param test_mode: Are we running in test mode?
        """
        if test_mode:
            return f"tests.{self.module_name}"
        else:
            return self.module_name


def merge_config(config, data):
    """Merge a new set of configuration requirements into a base configuration.

    :param config: the base configuration to update. This configuration is modified in-
        situ.
    :param data: The new configuration data to merge into the configuration.
    """
    # Properties that are cumulative lists
    for option in [
        "requires",
        "sources",
        "test_requires",
        "test_sources",
    ]:
        value = data.pop(option, [])

        if value:
            config.setdefault(option, []).extend(value)

    # Properties that are cumulative tables
    for option in ["permission"]:
        value = data.pop(option, {})

        if value:
            config.setdefault(option, {}).update(value)

    config.update(data)


def merge_pep621_config(global_config, pep621_config):
    """Merge a PEP621 configuration into a Briefcase configuration."""

    def maybe_update(field, *project_fields):
        # If there's an existing key in the Briefcase config, it takes priority.
        if field in global_config:
            return

        # Traverse the fields in the pep621 config; if the config exists, set it
        # in the Briefcase config.
        datum = pep621_config
        try:
            for key in project_fields:
                datum = datum[key]
        except KeyError:
            pass
        else:
            global_config[field] = datum

    # Keys that map directly
    maybe_update("description", "description")
    maybe_update("license", "license")
    maybe_update("url", "urls", "Homepage")
    maybe_update("version", "version")

    # Use the details of the first author as the Briefcase author.
    if "author" not in global_config:
        try:
            global_config["author"] = pep621_config["authors"][0]["name"]
        except (KeyError, IndexError):
            pass
    if "author_email" not in global_config:
        try:
            global_config["author_email"] = pep621_config["authors"][0]["email"]
        except (KeyError, IndexError):
            pass

    # Briefcase requires is cumulative over PEP621 dependencies
    try:
        pep621_dependencies = pep621_config["dependencies"]
        requires = global_config.get("requires", [])
        global_config["requires"] = pep621_dependencies + requires
    except KeyError:
        pass

    # Briefcase test_requires is cumulative over PEP621 optional test dependencies
    try:
        pep621_test_dependencies = pep621_config["optional-dependencies"]["test"]
        test_requires = global_config.get("test_requires", [])
        global_config["test_requires"] = pep621_test_dependencies + test_requires
    except KeyError:
        pass


def parse_config(config_file, platform, output_format, logger):
    """Parse the briefcase section of the pyproject.toml configuration file.

    This method only does basic structural parsing of the TOML, looking for,
    at a minimum, a ``[tool.briefcase.app.<appname>]`` section declaring the
    existence of a single app. It will also search for:

      * ``[tool.briefcase]`` - global briefcase settings
      * ``[tool.briefcase.app.<appname>]`` - settings specific to the app
      * ``[tool.briefcase.app.<appname>.<platform>]`` - settings specific to
        the platform
      * ``[tool.briefcase.app.<appname>.<platform>.<format>]`` - settings
        specific to the output format

    A configuration can define multiple apps; the final output is the merged
    content of the global, app, platform and output format settings
    for each app, with output format definitions taking precedence over
    platform, over app-level, over global. The final result is a single
    (mostly) flat dictionary for each app.

    :param config_file: A file-like object containing TOML to be parsed.
    :param platform: The platform being targeted
    :param output_format: The output format
    :returns: A dictionary of configuration data. The top level dictionary is
        keyed by the names of the apps that are declared; each value is
        itself the configuration data merged from global, app, platform and
        format definitions.
    """
    try:
        pyproject = tomllib.load(config_file)
    except tomllib.TOMLDecodeError as e:
        raise BriefcaseConfigError(f"Invalid pyproject.toml: {e}") from e

    try:
        global_config = pyproject["tool"]["briefcase"]
    except KeyError as e:
        raise BriefcaseConfigError("No tool.briefcase section in pyproject.toml") from e

    # Merge the PEP621 configuration (if it exists)
    try:
        merge_pep621_config(global_config, pyproject["project"])
    except KeyError:
        pass

    # For consistent results, sort the platforms and formats
    all_platforms = sorted(get_platforms().keys())
    all_formats = sorted(get_output_formats(platform).keys())

    try:
        all_apps = global_config.pop("app")
    except KeyError as e:
        raise BriefcaseConfigError("No Briefcase apps defined in pyproject.toml") from e

    # Build the flat configuration for each app,
    # based on the requested platform and output format
    app_configs = {}
    for app_name, app_data in all_apps.items():
        # At this point, the base configuration will contain a section
        # for each configured platform. Iterate over all the known platforms,
        # and remove these platform configurations. Keep a copy of the platform
        # configuration if it matches the requested platform, and merge it
        # into the app's configuration
        platform_data = None
        for p in all_platforms:
            try:
                platform_block = app_data.pop(p)

                if p == platform:
                    # If the platform matches the requested format, preserve
                    # it for later use.
                    platform_data = platform_block
                    merge_config(platform_data, platform_data)

                    # The platform configuration will contain a section
                    # for each configured output format. Iterate over all
                    # the output format of the platform, and remove these
                    # output format configurations. Keep a copy of the output
                    # format configuration if it matches the requested output
                    # format, and merge it into the platform's configuration.
                    format_data = None
                    for f in all_formats:
                        try:
                            format_block = platform_data.pop(f)

                            if f == output_format:
                                # If the output format matches the requested
                                # one, preserve it.
                                format_data = format_block

                        except KeyError:
                            pass

                    # If we found a specific configuration for the requested
                    # output format, add it to the platform configuration,
                    # overwriting any platform-level settings with format-level
                    # values.
                    if format_data:
                        merge_config(platform_data, format_data)

            except KeyError:
                pass

        # Now construct the final configuration.
        # First, convert the requirement definition at the global level
        merge_config(global_config, global_config)

        # The app's config starts as a copy of the base briefcase configuration.
        config = copy.deepcopy(global_config)

        # The app name is both the key, and a property of the configuration
        config["app_name"] = app_name

        # Merge the app-specific requirements
        merge_config(config, app_data)

        # If there is platform-specific configuration, merge the requirements,
        # then overwrite the platform-specific values.
        # This will already include any format-specific configuration.
        if platform_data:
            merge_config(config, platform_data)

        # Construct a configuration object, and add it to the list
        # of configurations that are being handled.
        app_configs[app_name] = config

    old_license_format = False
    for config in [global_config, *app_configs.values()]:
        if isinstance(config.get("license"), str):
            config["license"] = {"file": "LICENSE"}
            old_license_format = True

    if old_license_format:
        logger.warning(
            "Your app configuration has a `license` field that is specified as a string. "
            "Briefcase now uses PEP 621 format for license definitions. To silence this "
            'warning, replace the `license` declaration with `license.file = "LICENSE".'
        )

    return global_config, app_configs
